rbx_binary/src/deserializer/state.rs

```diff
@@ -85,6 +85,132 @@ struct Instance {
     children: Vec<i32>,
 }
 
+/// Properties may be serialized under different names or types than
+/// they ultimately should have in the DOM. CanonicalProperty
+/// represents the "proper" name and type of a property, and possibly
+/// contains a migration for some properties Roblox has replaced with
+/// others (like Font, which has been superceded by FontFace).
+#[derive(Debug)]
+struct CanonicalProperty<'a> {
+    name: &'a str,
+    ty: VariantType,
+    migration: Option<&'a PropertySerialization<'a>>,
+}
+
+fn find_canonical_property<'de>(
+    database: &'de ReflectionDatabase,
+    binary_type: Type,
+    class_name: &str,
+    prop_name: &'de str,
+) -> Option<CanonicalProperty<'de>> {
+    match find_property_descriptors(database, class_name, prop_name) {
+        Some(descriptors) => {
+            // If this descriptor is known but wasn't supposed to be
+            // serialized, we should skip it.
+            //
+            // On 2021-09-07 (v494), BasePart.MaterialVariant was added as a
+            // serializable Referent property. It was removed soon after, on
+            // 2021-10-12 (v499). Any models saved during that period have
+            // BasePart.MaterialVariant present.
+            //
+            // On 2022-03-08 (v517), BasePart.MaterialVariant was
+            // reintroduced as a string property that does not serialize. It
+            // serializes as MaterialVariantSerialized.
+            //
+            // In case we run into a model serialized during that period, or
+            // this happens again, we need to make sure that the name we
+            // found is the one that's supposed to serialize.
+            if let PropertyKind::Canonical { serialization } = &descriptors.canonical.kind {
+                if matches!(serialization, PropertySerialization::DoesNotSerialize) {
+                    log::debug!(
+                        "Skipping property {} as it is canonical and should not serialize.",
+                        descriptors.canonical.name
+                    );
+                    return None;
+                }
+            }
+
+            // TODO: Do we need an additional fix here?
+            let canonical_name = &descriptors.canonical.name;
+            let canonical_type = match &descriptors.canonical.data_type {
+                DataType::Value(ty) => *ty,
+                DataType::Enum(_) => VariantType::Enum,
+                _ => {
+                    // TODO: Configurable handling of unknown types?
+                    return None;
+                }
+            };
```

Comment:
> I realize these are copied verbatim from the old source but it does make me wonder what was actually meant by the first one.
> 
> With regards to the second one, we don't really need to worry about it because I don't foresee a world where Roblox adds something that's not a data type or an enum that also serializes.

These comments were added by https://github.com/rojo-rbx/rbx-dom/pull/230. 

The first comment might be talking about other possible ways rbx_binary can serialize properties that aren't supposed to serialize? But I'm not sure what other cases could lead to this kind of problem :shrug:

The fall through is hit on an unknown property type, and the second comment is referring to doing something smarter in that case. I think we should keep this comment.